### ISSUE TITLE
In view of the Angular Signals

### ISSUE THREAD

**Issue by Adam16** [state-adapt/state-adapt#31]
Hi,

I like State Adopt and I like to use it, also thanks to your videos on youtube ;-)
Just out of curiosity, I'd like to ask you about the position of State Adopt in relation to the Angular Signals? 

I'm interested in your opinion. Will this affect your plans for State Adopt?

Greeting from Germany.
Adam

**Comment by mfp22** [comment]
StateAdapt will get slightly easier to consume in components, but I don't see much changing yet. Someday I'll explore if signals can be used instead of selectors, but they can't in SolidJS so they probably can't in Angular either, but that's something I'll explore after I get experience with Angular signals.

For now, I'm thinking of creating utilities similar to what I have in SolidJS https://state-adapt.github.io/solid-js
But the naming will probably be more like what I have in React https://state-adapt.github.io/react

The only real question is what to name the signals, because I can't split it into a tuple. Maybe signals only, but a store property with the full store object? I don't know.

**Comment by mfp22** [comment]
Alex from the Angular team has explained to me that signals do not track their dependencies, so a lazy subscription inside `fromObservable` is not going to happen. Alex said it should be seen as similar to the async pipe. So the moment you are using `fromObservable`, you are saying, "I want this data."

So that confirms for me that the role that signals play in StateAdapt is the same as I said in the previous comment.

**Comment by Adam16** [comment]
Thanks for your detailed information and good luck ;-)

**Comment by mfp22** [comment]
Another update.

I had an idea last week that could be awesome.

When `adapt` is called, I can create a single signal for state and computeds for every selector. These will read state and call the selector function. But they will not use toSignal, so there will be no subscriptions created. Then, I can create an `injectStore` function that subscribes to `state$` internally and sets the state signal with the result. All the computeds pass state() into the selector functions. But only when they're read will they call the selector function.

So you can have this:

```typescript
const store = adapt(5, {
  selectors: {
    double: s => s * 2,
  }
});

signals = injectStates(store)
```

```html
<div>{{signals.state()}} doubled is {{signals.double()}}</div>
```

Names will change.

Injecting subscribes to state, but selectors are only evaluated when the computeds are needed.

**Comment by mfp22** [comment]
So, `injectStates` probably isn't the best name. Probably going to create `useAdapt`, which will combine `adapt` and `useStore`. Since we're not just injecting, but subscribing, `use` seems better: https://twitter.com/mfpears/status/1717245829931848175

Also looks like the syntax will be to add the signals as properties on the same store object, but prefix selector names with `$` to access the signal. https://twitter.com/mfpears/status/1716913120898335123

Doing this will not subscribe to the state (since you can read signals from anywhere) so the overall syntax will probably look something like this:
```typescript
@Component({
...
})
export class CountComponent {
  count = useAdapt(0, countAdapter); // Creates subscription immediately
}
```
```html
<p>Current count: {{ count.$state() }}</p>
<button (click)="count.increment()">Increment</button
```

Or

```typescript
@Injectable({ providedIn: 'root' })
export class CountService {
  store = adapt(0, countAdapter);
}

@Component({
...
})
export class CountComponent {
  count = useStore(inject(CountService).store); // Creates subscription now; unsubscribes when with `onDestroy`
}
```
```html
<p>Current count: {{ count.$state() }}</p>
<button (click)="count.increment()">Increment</button
```

Still thinking through some things. Is this `useStore` syntax annoying with the `useStore(inject(...).store);` double function call?

I just thought of doing something like Angular did with signals, where `signal` returns a function with properties. 

```typescript
count = signal(0);
constructor() {
  this.count() // call as function to get value
  this.count.set(n => n + 1); // Call property of signal as function
}
```

Maybe I could return a function from `adapt` that has all the normal properties stores have had, but when you call it, it activates it (creates a subscription). 

```typescript
@Component({
...
})
export class CountComponent {
  count = adapt(0, countAdapter)(); // Creates subscription immediately
}
```
```html
<p>Current count: {{ count.$state() }}</p>
<button (click)="count.increment()">Increment</button
```

Or

```typescript
@Injectable({ providedIn: 'root' })
export class CountService {
  store = adapt(0, countAdapter);
}

@Component({
...
})
export class CountComponent {
  count = inject(CountService).store(); // Creates subscription now; unsubscribes when with `onDestroy`
}
```
```html
<p>Current count: {{ count.$state() }}</p>
<button (click)="count.increment()">Increment</button
```

That might be complicated to implement, but it wouldn't require adding 2 more functions `useStore` and `useAdapt`. 

Maybe the React implementation could also change to this...

```tsx
export function Counter() {
  const [count, countStore] = adapt(0, countAdapter)();
  // Or maybe something like const [count, countStore] = adapt(0, countAdapter).$use();
  return ...
}
```

I don't know. It would be nice to not maintain 3 separate functions. 

Another question is what to do about signals that combine state from multiple stores. With `joinStores`, subscriptions are propagated up to each store upstream. Maybe there could be an Angular-specific utility like StateAdapt used to have. 

```typescript
// Old joinSelectors function
totalCount$ = joinSelectors(
  this.count1, // Uses `state` selector from `count1` store
  [this.count2, 'double'], // specify a selector
  (count1, count2) => count1 + count2,
);
```

The drawback of this was its unique syntax that made refactoring to `joinAdapters` rather than `joinStores` difficult. 

However, as time as gone on, I have seen not much need for this syntactic freedom. If stores are too granular, you can pass other store state as event context. Generally there has been no need to move from `joinStores` to `joinAdapters`. So maybe it's time to rethink some minimal syntax for derived states, and be okay with using signals for it. 

Nothing beats signals though... no need for a "dependency array". I wish we could just use signals:

```typescript
export class CombinedStateService {
  store1 = adapt(0);
  store2 = adapt(10);

  total = computed(() => this.store1.$state() + this.store2.$state());
}
```

Using `total` would not result in a subscription to the store. 

If I borrow from my "signal operator" or "signaler" idea, however...

```typescript
export class CombinedStateService {
  store1 = adapt(0);
  store2 = adapt(10);

  signalTotal = () => {
    const store1 = this.store1();
    const store2 = this.store2();
    return computed(() => store1.$state() + store2.$state());
  }
  signalTax = () => {
    const total = signalTotal();
    return computed(() => total() * 200_000);
  }
}

@Component({
...
})
export class CombinedStateComponent {
  tax = inject(CombinedStateService).signalTax(); // Subscriptions get created now
}
```

This doesn't seem better than `joinStores`. You can't create subscriptions inside `computed`s, so you end up with something kind of similar to dependency arrays anyway. Maybe this is okay? Let's compare to `joinStores`:

```typescript
export class CombinedStateService {
  store1 = adapt(0);
  store2 = adapt(10);
  
  combinedStore = joinStores({
    store1: this.store1,
    store2: this.store2,
  })({
    total: s => s.store1 + s.store2,
  })({
    tax: s => s.total * 200_000,
  })();
}

@Component({
...
})
export class CombinedStateComponent {
  tax = inject(CombinedStateService).combinedStore().$tax; // Subscriptions get created now
}
```

If it's between these 2 options, I do prefer this, but it's a bummer we can't start it off like 

```typescript
  combinedStore = joinStores({ store1, store2 })({
```

I guess I'm satisfied with `joinStores` though. I just don't know how to read a signal and propagate a subscription to a store without wrapping it in something that will be similar to defining a dependency array. And if you do need a dependency array, I can't think of syntax less boilerplatey than `joinStores`. 

Lots to think about still. Feedback would be appreciated.

**Comment by andsilmar** [comment]
Hi Mike, 

My humble opinion is I like the freedom to choose to use observables or signals for the same selector like the new `rxState` function of RxAngular using `state.select('property')` or `state.signal('property')`:

```typescript
// Definition example
interface State {
  pending: boolean;
  hasFailed?: boolean;
}

// with RxAngular
const state1 = rxState<State>(({ set }) => set({ pending: false}));
const pending1$ = state1.select('pending');
const pending1: Signal<boolean> = state1.signal('pending'); 
const hasFailed$ = state1.select('hasFailed');
const hasFailed: Signal<boolean | undefined> = state1.signal('hasFailed');

// with StateAdapt
const initialState: State = { pending: false }
const state2 = adapt(initialState, {
    selectors: {
        pending: s => s.pending,
        hasFailed: s => s.hasFailed
    }
})
const pending2$ = state2.pending$;
const pending2: Signal<boolean> = state2.pending; // <- I would like to have this in StateAdapt :)
```

Regards

**Comment by mfp22** [comment]
Of course you'll be able to choose. I think it will be like

```typescript
const pending2$ = state2.pending$; // observable
const pending2 = state2.$pending; // signal 
```

**Comment by timsar2** [comment]
How about signal selector in state adapt, Like ngrx signal store?

**Comment by Ban117** [comment]
Hey Mike, any updates on how state adapt will integrate with signals? Are you waiting for signal components/signal change detection?

**Comment by mfp22** [comment]
> Hey Mike, any updates on how state adapt will integrate with signals? Are you waiting for signal components/signal change detection?

Every time I think I have a plan, I have more ideas. That's the real problem. I don't want to commit to something and then make breaking changes. My job is pretty much to help companies upgrade dependencies, so I'm sensitive to this issue, but all library authors should be. That's why I created a script to automatically migrate from StateAdapt 1.x syntax to 2.x. 

I don't love the syntax of `$` prefix for signals. It's fine... but not amazing. So I'm thinking about that. 

Up to now the only way I've recommended using signals is to consume stores in templates. So a signal API doesn't matter much to me, and people haven't been bothering me for one either. In the mean time, they have to have an extra line of code for `toSignal` in their component. That's not great, but the lack of automatically generated signals isn't going to affect code structure in projects at this time. I have had a long time to consider the impact of signals on state management, and I still am not satisfied. For shared states, I've recommended using selectors. And this doesn't feel totally right. I'm trying to come up with a pattern that would allow signals to be used. 

But then the proposal to put signals in the web platform was released, and now I'm wondering if this is a great opportunity to rethink selectors, as well as signals. Because it appears to me that there's a way to determine if a signal is being used, and perform side effects when it's being used vs not used, which sounds perfect for managing RxJS subscriptions... But I have yet to dive deep into this. 

But most importantly, I have been traveling a lot and working full time, so I've really only had time to keep up with issues and should have a patch release coming soon, but no major features yet. There's a high chance I'll have a solid month to finally get to this stuff pretty soon. 

To summarize: I'm afraid of causing breaking changes, I haven't seen super high demand, there's a lot of stuff still evolving regarding signals, and I haven't had a ton of time.

**Comment by mfp22** [comment]
I'm going to replace selectors with signals. I will keep the syntax the same by default, but introduce a function for each framework library (like `@state-adapt/angular`) that will enable defining signals in that library's syntax. By default it will still use the `s => s.selector1 + s.selector2` syntax. But I have been bothered for a long time that 

```typescript
a = signal(0);
b = signal(10);
c = computed(() => this.a() + this.b());
```

is fine syntax, but then moving this to StateAdapt with a shared `c` selector would require refactoring:

```diff
a = adapt(0);
b = adapt(10);
c = joinStores({
  a: this.a,
  b: this.b,
})({
-c = computed(() => this.a() + this.b());
+  c: s => s.a + s.b,
});
```

The more computeds, the more refactoring.

The answer to this was, "Only use `computed` for template concerns. Keep it inside components." But I know that nobody will be perfect at following this. 

Classes and computeds are not as minimal as StateAdapt selectors. I refactored to the new potential syntax in one example and lines of code went from 154 to 177. But the benefit of consistency is very high. 

The tentative plan is first replace the internal implementation of selectors with the signals standard code, and deprecate `selectors` in favor of `signals: () => newSignals`. I need to remove my own auto-tracking selectors code. It's neat, but pretty complicated (especially the `joinStores` stuff) and still not quite as efficient as signals. This will also remove them from Redux devtools, but... there are people out there working on signals devtools, and I think that will be even better anyway.

Then I will export a specific function for each library like `@state-adapt/angular` to define derived state in state adapters with syntax that matches each library. For Angular it would be class syntax:

```typescript
import { buildAdapter } from '@state-adapt/angular';
// instead of @state-adapt/core, if I can figure out how to prevent imports from core

const adapter = buildAdapter<State>()({
  // state changes
})(S => class extends S {
  // Yeah, I've never seen anyone do this before, but it should totally work and be nice
})();
```

There's actually a lot more to think about still. I'll add details as I figure them out.

**Comment by mfp22** [comment]
Update: Signal support is probably finally coming soon. I need feedback on the ideas, and when released in StateAdapt 4.0 I will want feedback on the implementation.

## Why It Took So Long

This has been my primary frustration for years, and the implementation will be complicated because the Angular team refuses on principle to give us what we need here. The TC39 signals proposal has _exactly_ what we need, which is ironic, considering that it was copied from the Angular repo. Indeed, Angular itself has what I need, but it's deliberately hidden behind a private API. I technically can use it, but it exposes us to risk - although I may use it as a performance optimization, which the real durable solution being a fallback in case they change it.

When the TC39 proposal had a lot of momentum again, I finally looked into all of this and expected Angular to be forced to adhere to the common contract. Most issues on the TC39 repo that mention `watched` (what we need) simply assume it will be there when it lands. I had so much hope that Angular would finally support this.

I don't know what is going on anymore, but it's been too long and I have resigned myself to a messy implementation. The Angular team reaffirmed to me last week that they are still in principle opposed to signal reads causing "side-effects." 

StateAdapt is committed to total reactivity, not pure functional programming, and these things are often in direct opposition. So I don't care about the idealistic idea of reads being useless information. 

## Why This Matters

Some resources are expensive, and some reactive graphs are too complex for the end consumers at the bottom to have to reach up to try to co-manage resources at the top. I wrote about in the first half of [this article](https://dev.to/mfp22/introducing-the-auto-signal-pattern-1a5h)

The Angular team wants resources to be tied to explicit references, but if the lifetime is tied to an explicit reference, then anything shared between multiple components will be prematurely loaded in a common ancestor and not be able to manage itself efficiently when navigating between children of the common ancestor.

The injector tree is a tree, but data dependencies are a graph.

I once worked on a project with 100+ sibling routes. Many of them had common data sources. Some had 3-5 step waterfalls. It was practically unavoidable. So what happens? The parent of all 100+ routes got huuuuuuge so that on any given route, 90% of what you loaded was unnecessary. Also, since you needed Data D, you also had to dispatch an action to fetch Data A, B and C... or was it already in the store?

## TanStack Query vs Fine-Grained Reactivity

Meanwhile, this is trivial to manage in an actual reactive graph.

One approach is TanStack Query. It says "everything ask for what you want and we'll manage the central cache efficiently." It is aggressive with refetching - too aggressive imo. But it's amazing for the amount of care you no longer have to take. But what about derived states? I don't know what the solutions are now, but what I know from my experience in React is that you could share a useData hook, but anything derived from that has to be recomputed on behalf of every component that uses it. RTK has their own query integration so you could use selectors with that. It's not a _horrible_ issue - but it bothers me for the same reason React's rendering model will always bother me.

If signals could be reactive, that would be best. To be efficiently declarative/self-managing, a signal has to be able to tell a resource it represents "no need to refresh right now - nobody is listening" and "hey, someone is interested now - keep things accurate/fresh". The most harmless side-effect imaginable. It's not truly a side-effect much more than a pure function reserving memory for its variables. This is the same kind of affordance FRP opted for since the original "Functional Reactive Animations" paper - debounceTime is not a pure function, but purity and declarativeness/reactivity are in direct opposition.

## Strategy: Injector for Local, Ping for`providedIn: 'root'`

So, despite my persistent requests for official support, instead I have to figure out a nuanced way to mimic what I want as closely as possible.

I have a strategy that's a little complicated, and needs to be _thoroughly tested_, but I have it working locally right now.

1. If `ViewContainerRef` is present, the lifecycle of the store is simple and what most libraries have.

2. Otherwise, it's usually `providedIn: 'root'` and we take a different approach: When the store is inactive, upon the first read it wakes up. While it's active, every second it pings its graph and waits for a read from a `computed`. If it isn't read from, it tears itself down.

The one case where this could produce surprising behavior is for services that are not `providedIn: 'root'` and then are provided in the providers array of routes... They will still act as `providedIn: 'root'` because there isn't a `ViewContainerRef` in that environment. Not _super_ common, but I would argue in that case you should prefer `root`. However, I can of course support an option to make it behave the same way as if it was local. And I haven't thoroughly researched this case, but maybe there is a pretty simple thing to inject instead. Right now the router isn't technically a peer dependency, and handling this scenario directly would probably make it a peer dependency.

The truth is that in 99% of cases, root stores will only be torn down on route changes. And it actually is nice to keep the store alive for a second (configurable) in case the user navigates back, or there's a shared dependency on the new route. And I will allow an option to pass in an observable of navigation events instead, or any observable. But either way this is pretty complex and I am writing extremely thorough tests for it.

I wish I just had support for this from Angular instead. It's been my top request from the beginning. Maybe someday, official support will land. If it does, I can support an option to override this behavior and just make the activation/deactivation instant. Regardless, now that I have a clear answer from the Angular team, I am fully committed to figuring this out.

I downplayed it for a while, but this really will be nice:

```ts
count = adapt(0);
```
```html
<h2>{{ count() }}</h2>
```

No more `| async` or `toSignal`, _and_ no stupidly permanent stores. The _required_ pinging behavior is a minor inconvenience compared to not having basic reactive signal support.

**Comment by mfp22** [comment]
Well, this turned into an ordeal. After a lot of testing and learning, the approach above is not sophisticated enough. 

But, the new strategy will be much better. It's almost as good as if Angular supported this themselves, and it uses only stable APIs. And even if they changed the behavior of these stable APIs, what they are likely to move towards would likely be even better. And even if they choose worse API behavior, several backup strategies are possible. And it will be fully customizable - possibly even with the default strategy published as a standalone `toSignal` replacement. 

The biggest changes:
1. One-off reads will not activate stores. Instead, they will accumulate into a batch to be pinged later in the microtask queue to filter out passive from active consumers.
2. Stores will deactivate almost immediately once unused. The exact timing is still uncertain, but it will be no more than about 1 animation frame.

Stores created in components will still match the lifecycles of those components.

I'm excited to share more once I've finished implementing the new strategy and tested it thoroughly again.

**Comment by timsar2** [comment]
Tanstack query manage cache and working in ssr. I like to use state adapt but dont know if it support caching ssr,Also what is the plan of angular team for "resource api" manage and caching state in ssr.

**Comment by mfp22** [comment]
Finally done, will be released in StateAdapt 4.0 within a week.

Also exporting a `toSignal` replacement for Angular's imperative `toSignal`.

Early work
ae9573403f21aeb8f73709d1b88ca63f9609303b
5e678c8919a190cfa45aefe704f9066a22f7bac7
eedbe5d3f756b6e7e80f26e47aba5abf2880a797
f8d44a7e885a9bdbc08f4771e0a5b683b487ff8d

The good strategy:
49464b26803e74d9d07ceb302fdf3d594ae395ba
abea6ac86c64073df968b5a3c61fa469a526bf6f
